### ISSUE TITLE
Report question MADS

### ISSUE THREAD

**Issue by claralb2002** [MLOpsGroup101/StudentPerformanceML#90]
### Question 14

> **Upload 1 to 3 screenshots that show the experiments that you have done in W&B (or another experiment tracking**
> **service of your choice). This may include loss graphs, logged images, hyperparameter sweeps etc. You can take**
> **inspiration from [this figure](figures/wandb.png). Explain what metrics you are tracking and why they are**
> **important.**
>
> Recommended answer length: 200-300 words + 1 to 3 screenshots.
>
> Example:
> *As seen in the first image when have tracked ... and ... which both inform us about ... in our experiments.*
> *As seen in the second image we are also tracking ... and ...*
>
> Answer:

--- question 14 fill here ---

As seen in the attached screenshot, we tracked training loss, validation loss, and epochs. 

The Training Loss measures the Mean Squared Error on the training data after each epoch. This metric is fundamental as it quantifies how well the neural network is learning to map the input features (study habits, AI usage) to the target variable (student grades). The graph demonstrates a sharp decline in the first epoch, dropping from a loss of over 1800 to roughly 200. This rapid convergence indicates that the model architecture and learning rate are well-suited for the data, allowing the optimizer to quickly find a good set of weights.

The Validation Loss tracks the model's performance on a separate dataset that was held back during training. We track this to monitor for overfitting. It's common for a model to memorize training data, leading to low training loss but poor performance on new data. By comparing these two graphs, we can assess generalization. Since the validation loss trends downwards (from ~208 to ~190) and stabilizes alongside the training loss, we can confirm the model is generalizing well to unseen students.

We specifically chose MSE as our metric because this is a regression task. MSE is important because it penalizes large prediction errors more heavily than small ones, ensuring the model avoids making drastic mistakes in grade prediction. These metrics collectively inform us that 30 epochs were sufficient for the model to converge without overfitting.

### Question 15

> **Docker is an important tool for creating containerized applications. Explain how you used docker in your**
> **experiments/project? Include how you would run your docker images and include a link to one of your docker files.**
>
> Recommended answer length: 100-200 words.
>
> Example:
> *For our project we developed several images: one for training, inference and deployment. For example to run the*
> *training docker image: `docker run trainer:latest lr=1e-3 batch_size=64`. Link to docker file: <weblink>*
>
> Answer:

--- question 15 fill here ---

For our project we developed several docker images. We containerized each stage of the project using separate docker images including training, evaluation, inference, and the frontend. Docker images were based on the lightweight ghcr.io/astral-sh/uv:python3.13-bookworm-slim image. 

To for example build the training image, the following command would be used:

docker build -f dockerfiles/train.dockerfile -t stuperml-train'

We run the containers using volume mounts to persist artifacts (like trained models and figures) from the container back to the host machine, using the following command (it is also possible to change lr, amounts of epoch etc in the command):

docker run -v $(pwd)/models:/app/models stuperml-train

You can view the configuration of the train.dockerfile here: https://github.com/MLOpsGroup101/StudentPerformanceML/blob/7e8a05977ae496e252b2c0b97ed934ae2b56e5ec/dockerfiles/train.dockerfile

All containers were easily deployable in the GCP after testing locally.
For the API and Frontend, we configured the container to listen on port 8080 for Cloud Run deployment.
